### PR TITLE
Add boolean editor

### LINKED ISSUE #504: Editor for boolean type
We need to have editor for boolean type.
IMO:
- should be possible to switch between human-friendly editor and expressions (like in grafana)
- editor should be accessible for all fields with type: java.lang.Boolean, also for expressions in filters or switches
- editor should be by default in "expression" mode
- if expression saved in process have either true or false, we should render it using boolean editor. Otherwise (e.g. `#input.foo`) we should use "expression" mode
- default editor can be changed on backend side (will be done in other task)

### PR DESCRIPTION
**Pull request by PPiedel**
This resolves #504

### CODE REVIEW

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js [hunk 1]
```diff
@@ -57,37 +58,34 @@ export default class EdgeDetailsContent extends React.Component {
...
-                isMarked={this.isMarked("edgeType.condition.expression")}
-                showValidation={showValidation}
-              />
-            </div>
-          </div>
+          <EditableExpression
+            fieldType={"expression"}
+            fieldLabel={"Expression"}
+            renderFieldLabel={this.renderFieldLabel}
+            expressionObj={{expression: edge.edgeType.condition.expression, language: edge.edgeType.condition.language}}
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Maybe you can define expression object earlier?

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js [hunk 2]
```diff
@@ -57,37 +58,34 @@ export default class EdgeDetailsContent extends React.Component {
   }
 
   render() {
-    const { edge, readOnly, updateEdgeProp, showValidation} = this.props
+    const {edge, readOnly, updateEdgeProp, showValidation, showSwitch} = this.props
 
     switch (_.get(edge.edgeType, 'type')) {
       case "SwitchDefault": {
         return this.baseModalContent()
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Please do simple functional component: `const BaseModal Content = (props) => {}`. In props you should alsow have a children element for rendering it at NextSwitch.

File: ui/client/components/graph/node-modal/NodeDetailsContent.js [hunk 3]
```diff
@@ -439,43 +444,50 @@ export class NodeDetailsContent extends React.Component {
...
+          values={restriction.values}
+          isMarked={isMarked}
+          showSwitch={showSwitch}
+          readOnly={readOnly}
+          onValueChange={(newValue) => this.setNodeDataAt(exprTextPath, newValue)}
+        />
+      )
+
+    const param = this.findParamByName(fieldLabel)
+    return TestRenderUtils.wrapWithTestResult(
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/NodeDetailsContent.js]: This should be a separeted highlevel component. Now it looks ugly :/

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    super(props)
+    this.state = {
+      displayRawEditor: true
+    }
+  }
+
+  render() {
+    const {fieldType, expressionObj, rowClassName, valueClassName, showSwitch} = this.props
+    const editorName = resolveEditorName(fieldType, expressionObj, this.state.displayRawEditor)
+    const Editor = editorTypes[editorName]
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: What will happen when editorName not exists in editorTypes?

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      case rawEditor:
+        return parseableBoolean(expressionObj)
+      default:
+        return false
+    }
+  }
+
+  showSwitch = (fieldType, showSwitch) =>
+    showSwitch &&
+    (!(fieldType === "VariableBuilder" || fieldType === "Variable")
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: This strings should be defined as a const before componnet. Look at DeploymentAction in https://github.com/TouK/nussknacker/pull/518/files#diff-2bbdc336921dd7cd73e1f80fdff9f6b9R11

File: ui/client/components/graph/node-modal/editors/field/LabeledInput.js [hunk 6]
```diff
@@ -25,7 +25,7 @@ LabeledInput.propTypes = {
   renderFieldLabel: PropTypes.func.isRequired,
   isMarked: PropTypes.bool,
   readOnly: PropTypes.bool,
-  value: PropTypes.string,
+  value: PropTypes.string || PropTypes.number,
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/field/LabeledInput.js]: oO

**Comment by lciolecki** [comment]
Good job! It looks good but we should do some improvements.. :)

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js (hunk 4 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: resolveEditorName return by default rawEditor so this situation should not happen

File: ui/client/components/graph/node-modal/editors/field/LabeledInput.js (hunk 6 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/field/LabeledInput.js]: Well, it happens

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js (hunk 1 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Done

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js (hunk 2 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Done

File: ui/client/components/graph/node-modal/NodeDetailsContent.js (hunk 3 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/NodeDetailsContent.js]: Done, but not sure if we should refactor this in this PR

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js (hunk 5 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: done

File: ui/client/components/graph/TestRenderUtils.js [hunk 7]
```diff
@@ -52,7 +17,10 @@ export function testResults(nodeId, testResultsToShow) {
         {Object.keys(ctx).map((key, ikey) =>
           <div className="node-row" key={ikey}>
             <div className="node-label">{key}:</div>
-            {showTestValue(ctx[key], false)}
+            <TestValue
```
Comment (lciolecki) [inline ui/client/components/graph/TestRenderUtils.js]: It can be in one line :)

File: ui/client/components/graph/node-modal/BaseModalContent.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+import React from 'react'
+import PropTypes from "prop-types"
+import ModalRenderUtils from "./ModalRenderUtils"
+
+export default function BaseModalContent(props) {
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/BaseModalContent.js]: How beautilful.. :)

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js [hunk 9]
```diff
@@ -13,81 +13,66 @@ export default class EdgeDetailsContent extends React.Component {
...
+            expressionObj={expressionObj}
+            readOnly={readOnly}
+            validators={[notEmptyValidator]}
+            isMarked={this.isMarked("edgeType.condition.expression")}
+            showValidation={showValidation}
+            showSwitch={showSwitch}
+            onValueChange={(newValue) => updateEdgeProp("edgeType.condition.expression", newValue)}
+          />
+
+        return (
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Not in children.. But:
```
<BaseModalContent>
<EditableExpression />
</BaseModalContent>
```

File: ui/client/components/graph/node-modal/BaseModalContent.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  const {edge, edgeErrors, readOnly, isMarked, changeEdgeTypeValue, children} = props
+
+  return (
+    <div className="node-table">
+      {ModalRenderUtils.renderOtherErrors(edgeErrors, "Edge has errors")}
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/BaseModalContent.js]: This is alsow ugly way.. We shoudl have there another separeted component.. I mean all method from ModalRenderUtils should be in package modal/notifications/ Info / Warning / Error - alsow it looks like we can remove methods: renderWarning and renderInfo.

File: ui/client/components/graph/node-modal/NodeDetailsContent.js [hunk 11]
```diff
@@ -439,43 +445,55 @@ export class NodeDetailsContent extends React.Component {
     this.doCreateExpressionField(fieldName, fieldName, `${listFieldPath}.${expressionProperty}`, validators);
 
   doCreateExpressionField = (fieldName, fieldLabel, exprPath, validators) => {
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/NodeDetailsContent.js]: No you can do separeted component: `ExpressionField`. We can use it in future to refactor this huge NodeDetailsContent.

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  toggleEditor = (_) => this.setState({
+    displayRawEditor: !this.state.displayRawEditor
+  })
+
+  switchable = (editorName, expressionObj) => {
+    switch (editorName) {
+      case boolEditor:
+        return true
+      case rawEditor:
+        return parseableBoolean(expressionObj)
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: It's use only EditableExpression.js move it to this file ;)

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      default:
+        return false
+    }
+  }
+
+  showSwitch = (fieldType, showSwitch) => showSwitch && supportedFieldTypes.includes(fieldType)
+}
+
+const resolveEditorName = (fieldType, expressionObj, displayRawEditor) => {
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: Why you use once time stiing, other time defined value? There should be something like:

```
const EditorType = {
  BOOL_EDITOR: "boolEditor",
  RAW_EDITOR: "rawEditor",
 EXPRESSION: "expression",
 BOOLEAN: "Boolean",
 EXPRESSION_WITH_FIXED_VALUE: "expressionWithFixedValues",

supportedFieldTypes: ["expression", "Boolean"],

isSupported = (filedType) => this.supportedFieldTypes.includes(fieldType),

else methods correlated with fieldTypes.. 
}
```

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    case expressionWithFixedValuesEditor:
+      return fieldType
+    case rawEditor:
+      return fieldType
+    default:
+      return rawEditor
+  }
+}
+
+const editorTypes = {
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: I don't like this kind of getting from object.. We should always return some default value.. So my propose is move this to method in EditableExpression and use there switch with defaults.

File: ui/client/components/graph/node-modal/editors/map/MapValue.js [hunk 15]
```diff
@@ -1,23 +1,24 @@
...
-        }}
         validators={validators}
         isMarked={isMarked}
-        showValidation={showValidation}/>
+        readOnly={readOnly}
+        showValidation={showValidation}
+        showSwitch={showSwitch}
+        onValueChange={(value) => onChange(`${paths}.expression.expression`, value)}
+        expressionObj={value}
+        rowClassName={""}
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/map/MapValue.js]: Better way should be null / undefined -> without this props.

File: ui/client/components/graph/node-modal/BaseModalContent.js (hunk 10 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/BaseModalContent.js]: I agree that these piece of code need some refactoring. Think it should be done in separated PR.
What errors have in common with expression :P?

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js (hunk 12 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: Yes, but I think separation of responsibilities should be considered here - there will be more parse functions in the near future. I don't want to keep all these functions in React.Component

File: ui/client/components/graph/TestRenderUtils.js (hunk 7 above)
Comment (PPiedel) [reply ui/client/components/graph/TestRenderUtils.js]: Done

File: ui/client/components/graph/node-modal/EdgeDetailsContent.js (hunk 9 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/EdgeDetailsContent.js]: Done

File: ui/client/components/graph/node-modal/NodeDetailsContent.js (hunk 11 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/NodeDetailsContent.js]: Done

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js (hunk 13 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: Done

File: ui/client/components/graph/node-modal/editors/expression/EditableExpression.js (hunk 14 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/EditableExpression.js]: Done

File: ui/client/components/graph/node-modal/editors/map/MapValue.js (hunk 15 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/map/MapValue.js]: Done

File: ui/client/components/graph/node-modal/editors/expression/tests/TestValue.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+import React from "react"
+
+export default function TestValue(props) {
+
+  const {testValue, shouldHideTestResults} = props
+  const hiddenClassPart = (shouldHideTestResults ? " partly-hidden" : "")
+
+  return (
+    <div className={"node-value" + hiddenClassPart}>
+      {(testValue || {}).original ?
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/tests/TestValue.js]: Simple _.get() ? :)

Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/tests/TestValue.js]: Done

**Comment by witekw** [comment]
"Switch to more intuitive mode" is very colloquial, lets name it "basic mode" and "expression mode", what do you think?

**Comment by witekw** [comment]
Drop down list in a last element is not visible, it expands modal's content and reuires scrolling down to see it.

image

**Comment by witekw** [comment]
Switching to "more intuitive mode" button should be disabled (gray) when it is not possible to switch.

**Comment by witekw** [comment]
There are two ways of displaying dropdown lists near each other

image

**Comment by witekw** [comment]
The switch should not be there

image

**Comment by PPiedel** [comment]
> There are two ways of displaying dropdown lists near each other
> 
> image

Expression in that case depends on "condition" mode s I'm not sure if we can do something about that

**Comment by PPiedel** [comment]
> Drop down list in a last element is not visible, it expands modal's content and reuires scrolling down to see it.
> 
> image

I've extended edge details modal a bit

**Comment by PPiedel** [comment]
> The switch should not be there
> 
> image

Strange situation. It is displayed as a regular expression editor, but backend knows that it should be  string there. I've fixed it on the FE side

**Comment by PPiedel** [comment]
> "Switch to more intuitive mode" is very colloquial, lets name it "basic mode" and "expression mode", what do you think?

It,s ok, done

File: ui/client/components/graph/node-modal/editors/expression/ExpressionField.js [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+          readOnly={readOnly}
+          onValueChange={(newValue) => setNodeDataAt(exprTextPath, newValue)}
+        />
+      )
+
+    const param = this.findParamByName(fieldLabel)
+
+    const editableExpression =
+      <EditableExpression
+        fieldType={fieldType || (param ? ProcessUtils.humanReadableType(param.typ.refClazzName) : "expression")}
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]: This shoud be done in EditableExpression.

File: ui/client/components/graph/node-modal/editors/expression/ExpressionField.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        expressionObj={expressionObj}
+        validators={validators}
+        isMarked={marked}
+        showValidation={showValidation}
+        showSwitch={showSwitch}
+        readOnly={readOnly}
+        onValueChange={(newValue) => setNodeDataAt(exprTextPath, newValue)}
+      />
+
+    return (
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]:
```
<EditableExpressionWithTestResults>
<EditableExpression />
</EditableExpressionWithTestResults>
```

File: ui/client/components/graph/node-modal/editors/expression/ExpressionField.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+
+  getRestriction = (fieldName) => {
+    const restriction = (this.findParamByName(fieldName) || {}).restriction
+    return {
+      hasFixedValues: restriction && restriction.type === "FixedExpressionValues",
+      values: restriction && restriction.values
+    }
+  }
+
+  findParamByName(paramName) {
```
Comment (lciolecki) [inline ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]:
```
findParamByName = (paramName) => _.get(this.props, "nodeObjectDetails.parameters", [])).find((param) => param.name === paramName
```

Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]: Done

File: ui/client/components/graph/node-modal/editors/expression/ExpressionField.js (hunk 18 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]: Done

File: ui/client/components/graph/node-modal/editors/expression/ExpressionField.js (hunk 17 above)
Comment (PPiedel) [reply ui/client/components/graph/node-modal/editors/expression/ExpressionField.js]: Done

**Comment by witekw** [comment]
It looks fine now. I think it should be possible to switch to "basic mode" when the input is empty.

**Comment by PPiedel** [comment]
> One comment - allow basic mode when there is no value in the input.

Ok, done with default value: true

**Pull request merged**